hw/ip/otbn/util/shared/information_flow_analysis.py

```diff
@@ -0,0 +1,598 @@
+    # end of the loop like a RET instruction.
+    if loop_end_pc is not None and loop_end_pc in section:
+        assert loop_end_pc == section.end
+        loop_end_pc = None
+        edges = [Ret()]
+
+    # Get the information flow, used constants, and known constants at the end
+    # of the straightline block (not including the last instruction). Note that
+    # _build_iflow_straightline updates the `constants` dictionary in-place.
+    used_constants, iflow = _build_iflow_straightline(program, section.start,
+                                                      section.end - 4,
+                                                      constants)
+
+    # Get the instruction/operands at the very end of the block for special handling
+    last_insn = program.get_insn(section.end)
+    last_op_vals = program.get_operands(section.end)
+
+    last_insn_used_constants, last_insn_iflow = _build_iflow_insn(
+        last_insn, last_op_vals, section.end, constants)
+
+    # Update used constants to include last instruction
+    for const in last_insn_used_constants:
+        used_constants.update(iflow.sources(const))
+
+    # Update control_deps to include last instruction
+    last_insn_control_deps = _get_insn_control_deps(last_insn, last_op_vals)
+    for node in last_insn_control_deps:
+        for source_node in iflow.sources(node):
+            if source_node not in control_deps:
+                control_deps[source_node] = set()
+            control_deps[source_node].add(section.end)
+
+    # Update information-flow to include last instruction
+    iflow = iflow.seq(last_insn_iflow)
+
+    if last_insn.mnemonic == 'loopi' or last_insn.mnemonic == 'loop':
+        # Special handling for loops; reject non-constant #iterations, and run
+        # loop body #iterations times
+        iterations = _get_constant_loop_iterations(last_insn, last_op_vals,
+                                                   constants)
+        if iterations is None:
+            raise ValueError(
+                'LOOP instruction on a register that is not a '
+                'known constant at PC {:#x} (known constants: {}). If '
+                'the register is in fact a constant, you may need to '
+                'add constant-tracking support for more instructions.'.format(
+                    section.end, constants.keys()))
+
+        if len(edges) != 1 or not isinstance(edges[0], LoopStart):
+            raise RuntimeError(
+                'Control graph section ends in {} at PC {:#x} but the '
+                'next control-flow locations are: {} instead of 1 '
+                'LoopStart instance as expected'.format(
+                    last_insn.mnemonic, section.end, edges))
+        body_loc = edges[0]
+
+        # Update the constants to include the loop instruction
+        _update_constants_insn(last_insn, last_op_vals, constants)
+
+        # Recursive calls for each iteration
+        for _ in range(iterations):
```

Comment:
You probably know this already, but if you had some sort of "phi node support", you could avoid needing a constant number of iterations. It'd also be much faster if there were e.g. 1024 iterations to do! :-) (Clearly not for this PR, but I thought it was a good place to point it out)

Reply:
Yes, I do have a plan for this! It would also solve the problem with control-flow cycles that aren't loops and it would be cool. I thought I should probably get the easiest thing working first before getting too carried away, though :slightly_smiling_face: